### PR TITLE
Unit tests for NotifyVolunteerForUserUnenrolls #783

### PR DESCRIPTION
**Pull request by nedruk**
List of test cases:
- PreferredEmail is valid -> notification sent to preferred email.
- PreferredEmail is null -> notification sent to default email.
- PreferredEmail is empty -> notification sent to default email.
- PreferredEmail is white space -> notification sent to default email.
- default email is null -> notification not sent.
- default email is empty -> notification not sent.
- default email is white space -> notification not sent.
- volunteer is not signed up for event -> notification not sent.
- event is null -> notification not sent.
- event is empty -> notification not sent.

When notification is being sent, tests validate recipient, message subject and body. 

Let me know if it makes sense. 

Also, please note - "event is null" is currently failing.

Thanks

### CODE REVIEW

**Comment by nedruk** [comment]
Sorry, I don't understand why the build fails. 
Both VolunteerModel and TaskSummaryModel are from AllReady.Areas.Admin.Models namespace that is referenced at the top of the file.
What am I missing here?

**Comment by mgmccarthy** [comment]
@nedruk, I just pulled your branch down locally from this PR. I cleaned and rebuilt and it builds fine. However, when running all unit tests, the `NotifyVolunteerForUserUnenrolls_NotSent_EventIsNull()` unit test failed:
image

see if it fails for you locally. If it does, fix it and push again. Maybe that will allow the build to succeed.

**Comment by nedruk** [comment]
@mgmccarthy, I fixed up the failing test - didn't help, then commented it out all together. Build still fails. On my end 'clean-rebuild' works without issues. 
Looking at build output, looks like Web project is not being built prior to Test project. Classes that cause failure are from Web project. Could that be the problem?

**Comment by mgmccarthy** [comment]
@nedruk, I'

**Comment by mgmccarthy** [comment]
@nedruk, I'm not having any of the issues on any of my branches that I've submitted PR's for, so I'm not too sure how to replicate what you're seeing. 

Make sure you have latest from master and rebase. Once you've rebased and have handled your merge conflicts, push origin again and see if they build works. 

If you're not familiar with how to rebase for contributing to the AllReady project, please see Steve Gordon's blog post: http://stevejgordon.co.uk/contributing-to-allready

Again, this could be a build server issue, but I'm not a project owner, so I don't really have the ability to troubleshoot the build server in any way.

**Comment by MisterJames** [comment]
@nedruk First off, thanks for this PR! Looking forward to having you as a contributor on the project. Just had a quick review with @mgmccarthy and we have some feedback for you...he's going to fill you in. There are quite a few examples of our current MO here in the project, and unfortunately some that shouldn't be modeled after. We're going to help steer you in the right direction.  There will likely be many line-level comments, but please just take them as course-alignment.

Our number one priority is making sure that we have code that anyone else can pick up, so we try to keep things consistent and clear from class to class. Thanks again, and looking forward to seeing the recommended changes tracked in!

**Comment by mgmccarthy** [comment]
@nedruk, I want to work on a couple changes with you in order to get this PR into a "mergable" state along the lines of what @MisterJames was saying. At a high level, these are the things I'd like to work on with  you for the current PR

- the correct testing of Mediator
- the test class setup as well as population of un-needed properties
- test naming
- scoping tests to a single piece of functionality

I propose we work line by line through the `NotifyVolunteerForUserUnenrolls` handler, and eventually work our way to more granular tests. 

Let me know what your thoughts are. I could give you guidance in the form of code samples, or we could use this PR thread as a conversation thread to ask and answer questions.

Looking forward to it.

**Comment by nedruk** [comment]
Thanks @MisterJames and @mgmccarthy . I'll certainly be happy to learn how to do things the right way. :) 

@mgmccarthy , can we do a combination of the 2 options? Is it possible to make a bunch of comments right in code, marking what was done wrong and how you'd prefer me to do that? I understand that, given the comments above, there's going to be quite a few things.. After that, I'll start working on improvements and make clarification questions as I go in the PR thread.

Let me know your thoughts.

Thanks again for patience.

**Comment by mgmccarthy** [comment]
@nedruk, I'd like to take a different approach then throwing a ton of comments into the existing PR. 

I'd like to start with a fresh test, and to test the first "testable" thing in NotifyVolunteerForUserUnenrolls. Don't change anything about your current code you've written... let's add a fresh test to the top of the test class.

To start with, rename the test class from `NotifyVolunteerForUserUnenrollsTests` to `NotifyVolunteerForUserUnenrollsShould`.

Next, let's take a look at the the first couple lines of the Handle method:
image

what is going on here?
- EventDetailForNotificationQueryAsync is sent asyn via the mediator
- there is a model returned

seeing how we should only assert one thing (at least strive to do so) per test method, I want you to create a new test called:
```
[Fact]
public void SendEventDetailForNotificationQueryAsyncWithCorrectParameters()
{
}
```

I want you to aim for the **minimum setup required** for the test to pass. Minimum setup means the data populated on the classes involved in the test as well as the dependencies injected into the class. Just look at that line of code we want to test to determine what needs to be setup. 

Don't reuse any of your code you've already written that does shared setup. We want to move away from shared setup where ever we can unless it makes sense.

You can reply to this post with the test method or update this PR and I'll pull locally to check it out.

By working through a couple tests this way, you can start to get an idea of what we're after, and then go on the update the rest of the tests yourself.

Thanks! Looking forward to working through a couple of these with you.

**Comment by nedruk** [comment]
Got it. Will work on that one.
Thanks @mgmccarthy.

**Comment by nedruk** [comment]
Apologies for all those commits. It took a while to get my head around rebase-merge process. Hopefully it's all good now. 

@mgmccarthy, I've created the unit test trying to follow your recommendations. Below is the screenshot. Please have a look and let me know if any changes needed.

Thanks.

image

**Comment by mgmccarthy** [comment]
@nedruk, left a couple comments. If you have any questions about any of the corrections I mentioned, please ask me why I'm recommending them.

**Comment by nedruk** [comment]
Thanks again, @mgmccarthy . I've made amendments as per your recommendations. No questions from my side, they completely make sense.
 
Re this avalanche of others' commits upon rebase.. I read through that article as you'd advised earlier, and was following the cheat sheet. But I got a couple of conflicts on my way and apparently screwed up something while trying to resolve them.

I could try to put together another test case myself and see how it goes. What do you think? 

Thanks.

image

**Comment by mgmccarthy** [comment]
@nedruk, I find it useful to close the solution in visual studio before beginning a rebase. That way, you're not constantly being pestered to reload files into the IDE as rebase replays your changes. That way, you can use VS's comapre mechanisam to make the merge file be file as you go through the rebase.

Even doing all that, it's sometimes tough to make the right merge choices on rebase. There is one more article, and that was written by Dave Paquette on the whole rebase process for AllReady. You can read that blog post here: http://www.davepaquette.com/archive/2016/01/24/Submitting-Your-First-Pull-request.aspx

**Comment by nedruk** [comment]
Thanks for the detailed instructions and encouragement, @mgmccarthy.
Below is another bit of work.
Let me know if you are ok with it.

Thanks.

image

**Comment by nedruk** [comment]
@mgmccarthy, I've written 2 tests to verify that notification command:

- is sent when recipient email is provided (SendNotifyVolunteersCommand)
- is not sent when recipient email is unavailable (NotSendNotifyVolunteersCommand_WhenRecepientEmailIsNull) 

Please have a look.

Thanks.

image

**Comment by mgmccarthy** [comment]
@nedruk, you don't need the second test. As a matter of fact, the second test's name "SendNotificationVounteersCommand" doesn't really have any information in about **when** the handler is supposed to send NotificationVolunteersCommand. We'll get to the sending of the command soon after we get into the tests surrounding the StringBuilder and the message its building. You can delete the second test for now.

The first unit test has one more item in the setup that can be removed. Hint: check out your mediator Setup. Does code execute on both Setups given the path the conditional logic takes you down in the handler when you return from the handler?

**Comment by nedruk** [comment]
Updated both NotSendNotifyVolunteersCommand_.. tests.
@mgmccarthy , is that what you meant?

image

**Comment by mgmccarthy** [comment]
@nedruk, both of these tests are good.

However, when you look at the code that `NotSendNotifyVolunteersCommand_WhenEmailRecipientIsNull()` tests, you'll see that we can return from the handler for two reasons:
1. emailRecipient is null
2. emailRecipiient is "white space"

I'd create a new unit test called NotSendNotifyVolunteersCommand_WhenEmailRecipientIsWhiteSpace() which will be very similar to `NotSendNotifyVolunteersCommand_WhenEmailRecipientIsNull()`.

On a side note, it's a shame we have to make all these null and white space checks further down the MediatR message pipeline once the UI action that triggered the message(s) has already been completed. As you can see, it adds a lot of code to the handler and to unit tests.

What would have been better here is a decision up front that:
1. the Users email to be required at the point they fill it out on the UI
2. the PreferredEmail field is required when volunteering for an Event

or better yet, don't let a user use multiple emails for different events, make them supply an email to the site during registration, and until we have a confirmed email, do not let them volunteer for anything.

any of these decisions would make for much simpler processing once we leave the UI. But I don't control decisions like that ;)

**Comment by nedruk** [comment]
@mgmccarthy, done. 
I was thinking to also add tests cases for:
- mediator accepting EventDetailForNotificationQueryAsync returns null.
- model.UsersSignedUp list is not empty, but FirstOrDefault doesn't find a matching user.

Do you think those make sense?

**Comment by mgmccarthy** [comment]
@nedruk, last commit looks good. 

Also, nice catch on those other test cases.  They are both legit, so go ahead and write them. When you're done, we'll review.

Then we can move onto test the message construction and the command being sent.

**Comment by MisterJames** [comment]
@nedruk looking good so far. @mgmccarthy thanks for all the input and guidance here, this is great stuff (and a good PR that will serve as future reference). Thanks to both of you for the inline code and screen shots. 

@nedruk don't worry about the commit noise on this one...merge conflicts can be an uber pain in the buttocks, so let's chalk it up to learning exercises for now ;) I can "squash" your commits when we merge the final code in.

**Comment by nedruk** [comment]
@MisterJames, thanks for the encouraging words. :)

@mgmccarthy, I've added 2 more test cases, as discussed. Please have a look.
Currently, the code throws exception if the model returned by mediator is null. I've built the unit test around it. Is that the intended behaviour? 

Thanks. 

image

**Comment by mgmccarthy** [comment]
@nedruk, nice catch! 

Usually, if we do a read, and that read has the possibility of returning a null, the invoking code has some type of compensating action. That being said, we're not in a controller here, we're in a handler.

Also, we're in a handler doesn't do any type of "business" operation.. this handler's responsibility is to send a message about something that already happened in the system (in this case, a user has unenrolled)

For now, I would handle the null coming back from the `_mediator.SendAsync(new EventDetailForNotificationQueryAsync` call and return from the handler. 

```
var model = await _mediator.SendAsync(new EventDetailForNotificationQueryAsync { EventId = notification.EventId, UserId = notification.UserId })
                .ConfigureAwait(false);
if (model == null)
{
    return;
}
```

and then write the unit tests around that change.

the reason I'm saying this is because I don't want to think we'd want to throw an exception, which would end up rolling back the transaction that lets a user unenroll b/c our message handler failed. 

@MisterJames @tonysurma can a get a "yay" or "nay" on my statement I just made about this message handler?

Although MediatR is a a cool CQRS framework, it's not "fire and forget", which means the transaction of all messages send/consumed from an original message sent (usually from a controller) succeed or fail together.

**Comment by mgmccarthy** [comment]
@nedruk, any more unit test updates for this PR?

For now, let's assume the return; by checking if the model is null is something we want to change in the handler. After you change that and get the unit tests around that change, we can move onto the message construction tests along with the command being sent

**Comment by nedruk** [comment]
@mgmccarthy, i've rebased and made the recommended updates.

NotifyVolunteerForUserUnenrolls
image

NotifyVolunteerForUserUnenrollsShould
image

I'll work on unit tests around this line:
image

Will add 2 new tests to verify that SiteBaseUrl and EventId are populated correctly.

Sounds good?

**Comment by mgmccarthy** [comment]
@nedruk, awesome! 

Go ahead with those two new tests. I want you to start driving now based on the work we've done with this PR and I can move into doing the code reviews. Thanks for being patient and being open minded about hashing this out on GitHub with me. 

Open source needs more contributors like you!

**Comment by tonysurma** [comment]
Thank you @nedruk and @mgmccarthy the team work here is great to see and thank you both!

**Comment by nedruk** [comment]
Thanks for your kind words, @mgmccarthy, and for all your help making my first steps on GitHub.
Hope, this PR will be helpful for other newcomers.  ;)

File: AllReadyApp/Web-App/AllReady.UnitTest/Features/Notifications/NotifyVolunteerForUserUnenrollsTests.cs [hunk 1]
```diff
@@ -16,543 +16,336 @@ namespace AllReady.UnitTest.Features.Notifications
...
-        //            command = c;
-        //            return Task.FromResult(new Unit());
-        //        });
-
-        //    Should.Throw<NullReferenceException>(async() => await _handler.Handle(notification));            
-        //}
-
         [Fact]
-        public async Task NotifyVolunteerForUserUnenrolls_NotSent_EventIsEmpty()
+        public async Task SendNotifyVolunteersCommand_WithCorrectSiteBaseUrl()
```
Comment (mgmccarthy) [inline AllReadyApp/Web-App/AllReady.UnitTest/Features/Notifications/NotifyVolunteerForUserUnenrollsTests.cs]: you can delete this test (SendNotifyVolunteersCommand_WithCorrectSiteBaseUrl) and SendNotifyVolunteersCommand_WithCorrectEventId unit test and handle the construction of those two variables in the SendNotifyVolunteersCommand_WithCorrectEventLink unit test

**Comment by mgmccarthy** [comment]
@nedruk, for your latest commit, I left you one message in the PR changes itself. The tests you are writing are very very granular, and you're doing a little extra work you do not need to do.

For testing that the message is being constructed properly, you can probably combine all of that together into one test.

The message is constructed using these variables in the handler:
- `eventLink` (which is a result of string interpolation on some fixed text along with the `_options.Value.SiteBaseUrl` and `model.EventId` varaibles). This can be tested in the outgoing `NotifyVolunteersViewModel` class on the `NotifyVolunteersCommand` command
- `model.CampaignName`
- `model.EventName`

then you can assert that the constructed message matches the `EmailMessage` and `HtmlMessage` on `NotifyVolunteersViewModel`.

You can write separate tests to assert against `NotifyVolunteersViewModel.EmailRecipients` and `EmailRecipients.Subject`

**Comment by nedruk** [comment]
Thanks for the feedback, @mgmccarthy. I'm working on changes.

Re tests around NotifyVolunteersViewModel.EmailRecipients, I thought it was covered in SendNotifyVolunteersCommand_ToPreferredEmail and SendNotifyVolunteersCommand_ToUserEmail tests. Is there any scenario that I'm missing? Or is the naming wrong? 

Thanks.

File: AllReadyApp/Web-App/AllReady.UnitTest/Features/Notifications/NotifyVolunteerForUserUnenrollsTests.cs (hunk 1 above)
Comment (nedruk) [inline AllReadyApp/Web-App/AllReady.UnitTest/Features/Notifications/NotifyVolunteerForUserUnenrollsTests.cs]: @mgmccarthy , given we are consolidating all the tests asserting different parts of the message into one asserting the complete message, do we still need to test eventLink separately? It's included in the message body.

Comment (mgmccarthy) [inline AllReadyApp/Web-App/AllReady.UnitTest/Features/Notifications/NotifyVolunteerForUserUnenrollsTests.cs]: @nedruk, your are correct. We don't need to test eventLink separately.

**Comment by mgmccarthy** [comment]
@nedruk, you are correct. `SendNotifyVolunteersCommand_ToPreferredEmail` and `SendNotifyVolunteersCommand_ToUserEmail tests` does cover those tests are email recipients.

**Comment by nedruk** [comment]
@mgmccarthy, all done. Please have a look, and if all is good we could probably close this one. :)

**Comment by tonysurma** [comment]
thanks again @nedruk for working through all this.  @mgmccarthy if your final review is good we will merge

**Comment by mgmccarthy** [comment]
@tonysurma :shipit: 

@nedruk thanks so much! If you still have your original PR code laying around... take a look at how the tests were coded originally vs. what you ended up writing for it to be merged. There is a big difference. What we have now is something easier to understand, but more importantly (especially b/c this is open source)... it's something another developer can easily pick and work with.

**Pull request merged**